include/oneapi/dpl/internal/distributed_ranges_impl/detail/iterator_adaptor.hpp

```diff
@@ -0,0 +1,222 @@
+    }
+
+    bool
+    operator>(const_iterator other) const requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        return !(*this <= other);
+    }
+
+    bool
+    operator>=(const_iterator other) const requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        return !(*this < other);
+    }
+
+    reference
+    operator*() const
+    {
+        return *accessor_;
+    }
+
+    reference
+    operator[](difference_type offset) const
+        requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        return *(*this + offset);
+    }
+
+    iterator&
+    operator+=(difference_type offset) noexcept
+        requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        accessor_ += offset;
+        return *this;
+    }
+
+    iterator&
+    operator-=(difference_type offset) noexcept
+        requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        accessor_ += -offset;
+        return *this;
+    }
+
+    iterator
+    operator+(difference_type offset) const requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        iterator other = *this;
+        other += offset;
+        return other;
+    }
+
+    iterator
+    operator-(difference_type offset) const requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        iterator other = *this;
+        other += -offset;
+        return other;
+    }
+
+    difference_type
+    operator-(const_iterator other) const requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        return accessor_ - other.accessor_;
+    }
+
+    iterator&
+    operator++() noexcept requires(std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        *this += 1;
+        return *this;
+    }
+
+    iterator&
+    operator++() noexcept requires(!std::is_same_v<iterator_category, std::random_access_iterator_tag>)
+    {
+        ++accessor_;
+        return *this;
+    }
+
+    iterator
+    operator++(int) noexcept
+    {
+        iterator other = *this;
+        ++(*this);
+        return other;
+    }
+
+    iterator&
+    operator--() noexcept requires(std::is_same_v<iterator_category, std::random_access_iterator_tag> ||
+                                   std::is_same_v<iterator_category, std::bidirectional_iterator_tag>)
+    {
+        *this += -1;
+        return *this;
+    }
+
+    iterator
+    operator--(int) noexcept requires(std::is_same_v<iterator_category, std::random_access_iterator_tag> ||
+                                      std::is_same_v<iterator_category, std::bidirectional_iterator_tag>)
+    {
+        iterator other = *this;
+        --(*this);
+        return other;
+    }
+
+    auto
+    segments() const noexcept requires(ranges::__detail::has_segments_method<accessor_type>)
```

Comment:
What is the point of having `requires`-clause here. C++ doesn't case an error when users don't touch a function. The argument that concepts provide better diagnostics is very agruable.